presto-main/src/test/java/com/facebook/presto/operator/TableWriterOperatorTest.java

```diff
@@ -0,0 +1,130 @@
+import com.facebook.presto.spi.ConnectorOutputTableHandle;
+import com.facebook.presto.spi.ConnectorPageSink;
+import com.facebook.presto.spi.ConnectorPageSinkProvider;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.split.PageSinkManager;
+import com.facebook.presto.sql.planner.plan.TableWriterNode;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static com.facebook.presto.RowPagesBuilder.rowPagesBuilder;
+import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.testing.TestingTaskContext.createTaskContext;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static org.testng.Assert.assertEquals;
+
+public class TableWriterOperatorTest
+{
+    private static final String CONNECTOR_ID = "testConnectorId";
+    private final PageSinkManager pageSinkProvider = new PageSinkManager();
+    private final BlockingPageSink blockingPageSink = new BlockingPageSink();
+    private DriverContext driverContext;
+
+    @BeforeMethod
+    public void setUp()
+            throws Exception
+    {
+        ExecutorService executor = newCachedThreadPool(daemonThreadsNamed("test-%s"));
+        driverContext = createTaskContext(executor, TEST_SESSION)
+                .addPipelineContext(true, true)
+                .addDriverContext();
+
+        pageSinkProvider.addConnectorPageSinkProvider(CONNECTOR_ID, new BlockingConnectorPageSinkProvider(blockingPageSink));
+    }
+
+    @Test
+    public void testBlockedPageSink()
+            throws Exception
+    {
+        TableWriterOperator.TableWriterOperatorFactory factory = new TableWriterOperator.TableWriterOperatorFactory(
+                0,
+                pageSinkProvider,
+                new TableWriterNode.CreateHandle(new OutputTableHandle(CONNECTOR_ID, new ConnectorOutputTableHandle() {})),
+                ImmutableList.of(0),
+                Optional.empty(),
+                TEST_SESSION);
+
+        Operator operator = factory.createOperator(driverContext);
+
+        assertEquals(operator.isFinished(), false);
+        assertEquals(operator.needsInput(), true);
+
+        operator.addInput(rowPagesBuilder(BIGINT).row(1).build().get(0));
+
+        assertEquals(operator.isBlocked().isDone(), false);
+        assertEquals(operator.isFinished(), false);
+        assertEquals(operator.needsInput(), false);
+    }
+
+    private static class BlockingConnectorPageSinkProvider
```

Comment:
This is not Blocking, it just more SingleInstance ConnectorPageSinkProvider. It is quite general class which could be extracted from this class